File: src/main/java/com/hileco/cortex/primitives/layer/LayeredVmSet.java
package com.hileco.cortex.primitives.layer;

import com.hileco.vm.primitives.VmSet;

public class LayeredVmSet<T> implements VmSet<T, LayeredVmSet<T>> {
    private SetLayer<T> edge;

    public LayeredVmSet() {
        this(new SetLayer<>(null));
    }

    private LayeredVmSet(SetLayer<T> edge) {
        this.edge = edge;
    }

    public boolean contains(T key) {
        return contains(edge, key);
    }

    private boolean contains(SetLayer<T> layer, T key) {
        if (layer.getEntries().contains(key)) {
            return true;
        }
        if (layer.getDeletions().contains(key)) {
            return false;
        }
        var parent = layer.getParent();
        if (parent == null) {
            return false;
        }
        return contains(parent, key);
    }

    public void add(T key) {
        if (!contains(key)) {
            edge.getDeletions().remove(key);
            edge.getEntries().add(key);
        }
    }

    public void remove(T key) {
        if (contains(key)) {
            edge.getDeletions().add(key);
            edge.getEntries().remove(key);
        }
    }

    public int size() {
        throw new UnsupportedOperationException();
    }

    public LayeredVmSet<T> copy() {
        while (edge != null && edge.isEmpty()) {
            this.edge = edge.getParent();
        }
        var child1 = new SetLayer<>(edge);
        var child2 = new SetLayer<>(edge);
        this.edge = child1;
        return new LayeredVmSet<>(child2);
    }

    public void close() {
        edge.close();
    }
}

File: cortex-data/src/main/kotlin/com/hileco/cortex/data/layer/LayeredVmSet.kt
package com.hileco.cortex.data.layer

import com.hileco.cortex.data.VmSet

class LayeredVmSet<T> private constructor(private var edge: SetLayer<T>) : VmSet<T, LayeredVmSet<T>> {

    constructor() : this(SetLayer<T>(null))

    override fun contains(value: T): Boolean {
        return contains(edge, value)
    }

    private fun contains(layer: SetLayer<T>, value: T): Boolean {
        if (layer.entries.contains(value)) {
            return true
        }
        if (layer.deletions.contains(value)) {
            return false
        }
        val parent = layer.parent ?: return false
        return contains(parent, value)
    }

    override fun add(value: T) {
        if (!contains(value)) {
            edge.deletions.remove(value)
            edge.entries.add(value)
        }
    }

    override fun remove(value: T) {
        if (contains(value)) {
            edge.deletions.add(value)
            edge.entries.remove(value)
        }
    }

    override fun size(): Int {
        throw UnsupportedOperationException()
    }

    override fun copy(): LayeredVmSet<T> {
        while (edge.isEmpty) {
            this.edge = edge.parent ?: return LayeredVmSet()
        }
        val child1 = SetLayer(edge)
        val child2 = SetLayer(edge)
        this.edge = child1
        return LayeredVmSet(child2)
    }

    override fun close() {
        edge.close()
    }
}
